File: trees/bst/bst_iterator.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# Approach 1: Using shift and inorder traversal for one-time iteration.
class BSTIterator
    # @param {TreeNode} root
    def initialize(root)
        @list = inorder(root)
    end

    # @return {Boolean}
    def has_next
        !@list.empty?
    end

    # @return {Integer}
    def next
        @list.shift
    end

    private
    def inorder(root, visited = [])
        return [] if !root
        inorder(root.left, visited)
        visited << root.val
        inorder(root.right, visited)
    
        visited
    end
end

class BSTIterator
    def initialize(root)
        @stack = []
        push_root_to_left(root)
    end

    def next()
        root = @stack.pop
        push_root_to_left(root.right)
        root.val
    end

    def has_next()
        !@stack.empty?
    end
    
    private
    def push_root_to_left(root)
        while root
            @stack.push(root)
            root = root.left
        end
    end
end

# Your BSTIterator will be called like this:
# i, v = BSTIterator.new(root), []
# while i.has_next()
#    v << i.next
# end


# 173. Binary Search Tree Iterator
# https://leetcode.com/problems/binary-search-tree-iterator/

# Intuition
# 1. Iterating nodes of a BST is nothing but getting nodes in an in-order traversal order
# 2. We need some temp storage for nodes

# Approach 1: Simple list to store all elements (Sub-Optimal)
# Steps:
# 1. Save all elements in a list at init doing an inorder traversal, (left, root, right)
# 2. Keep popping from the list to provide for next.

# Time: next(): O(1), has_next(): O(1), init: ()
# Space: O(n)

# Approach 2: Using Stack
# Steps:
# 1. We keep a stack and at init we push the entire left sub-tree in reverse order
#    Like so, save root, root.left, root.left.left ... and so on until leaf node
# 2. When next is called, we pop the stack push the right sub-tree of the popped node
#    and return the value at the node
# 3. For hasNext() we just return if there are any nodes left in the stack
# 4. For push_left, we just iteratively push nodes into the stack
#    and move to the left node until the leaf

# Time: init: O(log(n)) cuz we only iterate down the left sub-tree
#       next(): O(1) per call, called n times
#       has_next(): O(1)
# Space: O(n)

File: trees/bst/bst_iterator.py
class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.stack = []
        self.push_left(root)

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return self.stack

    # @return an integer, the next smallest number
    def next(self):
        tmp = self.stack.pop()
        self.push_left(tmp.right)
        return tmp.val

    def push_left(self, node):
        while node:
            self.stack.append(node)
            node = node.left

# 173. Binary Search Tree Iterator
# https://leetcode.com/problems/binary-search-tree-iterator/

# I use Stack to store directed left children from root.
# When next() be called, I just pop one element and process its right child as new root.
# The code is pretty straightforward.

# So this can satisfy O(h) memory, hasNext() in O(1) time,
# But next() is O(h) time.

# I can't find a solution that can satisfy both next() in O(1) time, space in O(h).